eip_operator/src/controller/eip.rs

```diff
@@ -80,13 +93,207 @@ impl k8s_controller::Context for Context {
                     .public_ip
                     .as_ref()
                     .ok_or(Error::MissingPublicIp)?;
-                (allocation_id.to_owned(), public_ip.to_owned())
+                status.allocation_id = Some(allocation_id.clone());
+                status.public_ip_address = Some(public_ip.clone());
             }
             _ => {
                 return Err(Error::MultipleEipsTaggedForPod);
             }
         };
-        crate::eip::set_status_created(&eip_api, name, &allocation_id, &public_ip).await?;
+
+        // get potential attachments
+        let mut matched_pods: Vec<Pod> = vec![];
+        let mut matched_nodes: Vec<Node> = vec![];
+
+        match eip.spec.selector {
+            EipSelector::Pod { ref pod_name } => {
+                matched_pods = pod_api
+                    .list(
+                        &ListParams{field_selector: Some(format!("metadata.name={}", pod_name)), ..Default::default() }
+                    )
+                    .await?
+                    .into_iter()
+                    .filter(|pod| pod.metadata.deletion_timestamp.is_none())
+                    .collect::<Vec<Pod>>();
+                matched_pods.sort_unstable_by_key(|s| s.name_unchecked());
+            }
+            EipSelector::Node { ref selector } => {
+                let label_selector = selector
+                    .iter()
+                    .map(|(k, v)| format!("{}={}", k, v))
+                    .collect::<Vec<String>>()
+                    .join(",");
+                matched_nodes = node_api
+                    .list(
+                        &ListParams{ label_selector: Some(label_selector), ..Default::default()}
+                    )
+                    .await?
+                    .into_iter()
+                    .filter(|node| node.metadata().deletion_timestamp.is_none())
+                    .collect::<Vec<Node>>();
+                matched_nodes.sort_unstable_by_key(|s| s.name_unchecked());
+            }
+        }
+
+        // Check to make sure our resource still matches
+        // incase our selectors have updated, or the nodes/pods we have
+        // claimed have changed
+        // let correctly_associated =
+        let mut disassociate = false;
+        match &eip.spec.selector {
+            EipSelector::Pod { pod_name: _ } => {
+                if !matched_pods.is_empty() && eip.associated_with_pod(&matched_pods[0]) {
+                    status.claim = Some(matched_pods[0].name_unchecked());
```

Comment:
this is confusing to me - it feels like we shouldn't be writing to `status.claim` here at all, because the pod and node controllers are the ones managing the claims? like, it looks like the pod and node controllers are writing to `status.claim`, but then nothing actually reads that value - the eip controller just recalculates `status.claim` on its own. who should be the source of truth here?

Reply:
yeah I agree.. this is not great.
When a node/pod is modified, created, or destroyed we want to provide the opportunity for that resource to have it's EIP adjusted correctly without waiting for the EIP reconciliation to happen on its own schedule. 

We could have pods/nodes be the source of truth, but if we create an EIP we don't want it to have to wait for a pod/node to reconcile for it to assign itself. 

I have an idea that nodes/pods could just ask a particular EIP or set of EIPs they match to reconcile, but not modify the EIP at all.. I think we could get rid of claims then. It's not quite that straightforward, but I think it wouldn't be too complicated either.

Reply:
yeah, like, i think that is effectively what you have implemented here, just in a confusing way? like nothing ever actually reads from the eip.status.claims field, but by writing to it, you force an eip reconciliation, at which point the eip operator just recalculates the claims itself (the fact that they are stored in the status struct is kind of irrelevant because they are never used except further down in the same function)